src/Lecture2.hs

```diff
@@ -1,170 +1,292 @@
+--     able to continue the fight so they run away.
+--
+-- Implement data types to describe treasure, knight and dragon.
+-- And implement a function that takes a knight and a dragon and returns
+-- one of the three possible fight outcomes.
+--
+-- You're free to define any helper functions.
+--
+-- 🕯 HINT: If you find the description overwhelming to implement entirely
+--   from scratch, try modelling the problem in stages.
+--
+--     1. Implement all custom data types without using polymorphism.
+--     2. Add @newtype@s for safety where you think is appropriate.
+--     3. Encode the fight result as a sum type.
+--     4. Add polymorphism.
+--     5. Make all invalid states unrepresentable. Think, how you can
+--        change your types to prevent green dragons from having any
+--        treasure besides gold (if you already haven't done this).
 
 -- some help in the beginning ;)
-data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
-    }
 
-dragonFight = error "TODO"
+data Knight = Knight
+  { knightHealth :: Health,
+    knightAttack :: Attack,
+    knightEndurance :: Endurance
+  }
+
+data Dragon a
+  = RedDragon DragonStats (Chest a)
+  | BlackDragon DragonStats (Chest a)
+  | GreenDragon DragonStats Gold
+
+data DragonStats = DragonStats
+  { dragonStatsHealth :: Health,
+    dragonStatsAttack :: Attack
+  }
+
+dragonStats :: Dragon a -> DragonStats
+dragonStats (RedDragon stats _) = stats
+dragonStats (BlackDragon stats _) = stats
+dragonStats (GreenDragon stats _) = stats
+
+dragonHealth :: Dragon a -> Health
+dragonHealth = dragonStatsHealth . dragonStats
+
+dragonAttack :: Dragon a -> Attack
+dragonAttack = dragonStatsAttack . dragonStats
+
+setDragonStats :: DragonStats -> Dragon a -> Dragon a
+setDragonStats s d = case d of
+  RedDragon _ loot -> RedDragon s loot
+  BlackDragon _ loot -> BlackDragon s loot
+  GreenDragon _ loot -> GreenDragon s loot
+
+data Chest a = Chest
+  { chestGold :: Gold,
+    chestTreasure :: Treasure a
+  }
+
+data Treasure a = NoTreasure | SomeTreasure a
+
+data Reward a = Reward
+  { rewardExp :: Experience,
+    rewardGold :: Gold,
+    rewardTreasure :: Treasure a
+  }
+
+data FightResult a
+  = KnightWin (Reward a)
+  | KnightLose
+  | KnightFlee
+
+-- Newtypes to avoid ambiguities in the calculateDamage phase
+newtype Attack = Attack {unAttack :: Int}
+
+-- NOTE (Student) Deriving typeclasses makes things easier for calculations
+-- but we will use records members for now instead of
+-- deriving (Num, Eq, Ord)
+newtype Health = Health {unHealth :: Int}
+
+type Endurance = Int
+
+-- NOTE (Student) Ideally, we should represent the fact that
+-- gold is a positive int. I found the following answer from
+-- Gabriella Gonzalez https://stackoverflow.com/a/11910221
+-- However, it doesn't seem possible to prevent creating a Positive Int
+-- with a wrapped negative int if we are in the same module.
+type Gold = Int
```

Comment:
Indeed, the recommended approach might not be helpful in this module. In real world, it makes sense to create a `Positive` type in a separate module, don't export it's constructor but export a safe creating function. This pattern is known as the "Smart constructor" pattern in Haskell.

Haskell has several types for representing non-negative numbers. The most common types are:

* [Natural](https://hackage.haskell.org/package/base-4.18.0.0/docs/Numeric-Natural.html): unbounded, slow, subtraction throws a runtime error
* [Word](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Word.html): bounded, fast, overflows

You can choose one of them depending on the situation